doc/Precision_Guidelines_for_Ngen_Code_Validation.md

```diff
@@ -0,0 +1,36 @@
+
+In our tests of the Conceptual Functional Equivalent (CFE) versus CFE implemented via the Basic Model Interface (BMI) modularization methods, earlier comparisons checked six digits after the decimal points from the formatted outputs for catchment id-87. For these comparisons, we have seen identical values between the BMI-CFE and the original CFE code for Schaake Output Runoff, GIUH Runoff, Nash Lateral Flow, Ground Water Flow, and Total Discharge, for all 720 time steps, with 1 hour time step size. These tests are performed using varying initial conditions: soil_storage_percentage at 0.22, and 0.667; groundwater_storage_percentage at 0.5 and 0.01. We also looked at the total discharge data from standard screen output, which indicated 8-9 digits agreement between BMI-CFE and CFE calculations. In earlier tests, we have also shown similar agreement between CFE calculation results and the original T-Shirt C code calculation results.
+
+
+**Possible reasons for precision discrepancies between CFE and BMI-CFE:**
+
+The double precision numbers have an accuracy of 15 decimal digits, the loss of accuracy from rounding off error is expected, but precisely how many digit loss probably varies and depends on the specific code. Since BMI-CFE code has additional interface and difference in coding details, we hypothesize the memory allocation, cache usage, data sets loading into the registers may be different. These likely affect the rounding off processes and cause the least significant digits to differ. According to some literature, computer hardware architecture, compiler, even optimization may all affect numerical precision. 
+        
+In our CFE and BMI-CFE validation example, one particular case, deserves attention, which is: a key input variable, the rain rate, or specifically: precip_kg_per_m2, is a float point number, this could potentially reduces the precision in comparison between the two models even though all other variables are double, in consideration of rounding-off error in machine precision. There are many excellent discussions on rounding-off errors in literature due to machine precision, we refer interested readers to literature<sup>1-4</sup> for details. 
+
+On the other hand, the BMI-CFE and the original CFE demonstrate remarkable stability with 8-9 decimal point accuracy even after 720 time steps (hours). What are the possible reasons? We hypothesize two possible causes:
+1. The codes do not use space grids, all flow values are single valued for the catchment, i.e., there is no instability introduced by spatial matrices such as those in solving finite difference equations.
+2. Mass balance: In the evolution in time, the flows are constrained by mass balance. The mass balance, like energy conservation in classical molecular dynamics, should constrain the flow variation to a certain range, thus limiting the error propagation magnification in time (possibly through error cancellation).

```

Comment:
Did the test actually account for/accumulate mass balance error?

Reply:
Based on my experience with molecular dynamics modeling, energy conservation law ensures a few most significant digits is unchanged during time evolution. I believe mass balance does the same thing for flows in time evolution. It keeps the flow values from drifting away in one direction. A correct code should have this property. We did not specifically test the mass balance error. In theory, if the computer is infinitely accurate, mass balance should be strictly followed. My guess is that the mass balance error is random and related to round off error. For a stable algorithm, a few most significant digits should be unchanged during time evolution.